payments/api.py

```diff
@@ -476,6 +490,384 @@ def get_auto_apply_discounts_for_basket(basket_id: int) -> QuerySet[Discount]:
+                raise ValueError(integrated_system_missing_msg)  # noqa: B904
+    else:
+        integrated_system = None
+
+    if "product" in kwargs and kwargs["product"] is not None:
+        # Try to get the product via ID or SKU.  Raise an exception if it doesn't exist.
+        product_missing_msg = f"Product {kwargs['product']} does not exist."
+        if kwargs["product"].isdigit():
+            try:
+                product = Product.objects.get(pk=kwargs["product"])
+            except Product.DoesNotExist:
+                raise ValueError(product_missing_msg)  # noqa: B904
+        else:
+            try:
+                product = Product.objects.get(sku=kwargs["product"])
+            except Product.DoesNotExist:
+                raise ValueError(product_missing_msg)  # noqa: B904
+    else:
+        product = None
+
+    if "users" in kwargs and kwargs["users"] is not None:
+        # Try to get the users via ID or email.  Raise an exception if it doesn't exist.
+        users = []
+        user_missing_msg = "User %s does not exist."
+        for user_identifier in kwargs["users"]:
+            if user_identifier.isdigit():
+                try:
+                    users.append(User.objects.get(pk=user_identifier))
+                except User.DoesNotExist:
+                    raise ValueError(user_missing_msg % user_identifier)  # noqa: B904
+            else:
+                try:
+                    user = User.objects.get(email=user_identifier)
+                    users.append(user)
+                except User.DoesNotExist:
+                    raise ValueError(user_missing_msg % user_identifier)  # noqa: B904
+    else:
+        users = None
+
+    generated_codes = []
+
+    for code_to_generate in codes_to_generate:
+        with reversion.create_revision():
+            discount = Discount.objects.create(
+                discount_type=discount_type,
+                redemption_type=redemption_type,
+                payment_type=payment_type,
+                expiration_date=expiration_date,
+                activation_date=activation_date,
+                discount_code=code_to_generate,
+                amount=amount,
+                is_bulk=True,
+                integrated_system=integrated_system,
+                product=product,
+                bulk_discount_collection=bulk_discount_collection,
+            )
+        if users:
+            discount.assigned_users.set(users)
+
+        generated_codes.append(discount)
+
+    return generated_codes
+
+
+def update_discount_codes(**kwargs):  # noqa: C901, PLR0912, PLR0915
+    """
+    Update a discount code (or a batch of discount codes) as specified by the
+    arguments passed.
+
+    Keyword Args:
+    * discount_codes - list of discount codes to update
+    * discount_type - one of the valid discount types
+    * payment_type - one of the valid payment types
+    * redemption_type - one of the valid redemption types (overrules use of the flags)
```

Comment:
`redemption_type` is not exposed in the actual command, so this field can't be updated. Also, `one-time` and `one-time-per-user` are exposed in the command, but they're not checked here, so they don't work either.